src/persist-client/src/machine.rs

```diff
@@ -0,0 +1,400 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cmp;
+use std::collections::HashMap;
+use std::time::Instant;
+
+use anyhow::anyhow;
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use mz_persist_types::Codec64;
+use timely::progress::{Antichain, Timestamp};
+use timely::PartialOrder;
+
+use crate::collection::Collection;
+use crate::error::{InvalidUsage, LocationError};
+use crate::metadata::{CollectionMeta, ReaderMeta, TraceBatchMeta, WriterMeta};
+use crate::read::ReaderId;
+use crate::write::WriterId;
+use crate::{Log, SeqNo};
+
+#[derive(Clone, Debug)]
+pub struct ReadCapability<T> {
+    pub(crate) seqno: SeqNo,
```

Comment:
this sequence number (and the corresponding one for writers) is never modified at the moment. My understanding so far is:

- readers periodically need to say: I recognize a new meta at sequence number <foo> which is strictly greater than / equal to the meta I previously knew about <bar>

- as all readers advance their sequence numbers in this way, we can eventually delete meta's at old sequence numbers like <bar> once we know no reader can possibly point to them and then we can also remove old batches that are in <bar> but not in currently readable meta (aka meta at seqno_since)?

Is this understanding accurate?

Reply:
It seems like we want some kind of state machine input for readers to `downgrade_seqno(..)` or something like that - does that sound right? going to continue reading

Reply:
> My understanding so far is:
> 
> * readers periodically need to say: I recognize a new meta at sequence number  which is strictly greater than / equal to the meta I previously knew about
> * as all readers advance their sequence numbers in this way, we can eventually delete meta's at old sequence numbers like  once we know no reader can possibly point to them and then we can also remove old batches that are in  but not in currently readable meta (aka meta at seqno_since)?
> 
> Is this understanding accurate?

yes, that's exactly correct modulo one thing: as i have things written, the source of truth for which seqno a reader is on is now in the state machine state, not what the read handle thinks it is

> this sequence number (and the corresponding one for writers) is never modified at the moment. 

ah good catch. having this on the read handle now feels a bit vestigial from where this prototype started (implementing petros's design doc). it also makes my "modulo one thing" above feel like a footgun. feels like we want to do something different here than what I have but I'm not sure yet what it is. we could make what you said (readers explicitly forward it) work by e.g. returning a seqno from downgrade_since and then following it up with a second call to downgrade the seqno, but i'm already a bit worried about the volume of downgrade_since traffic and that just makes it worse. summary: dunno, more thinking needed here

Reply:
> It seems like we want some kind of state machine input for readers to downgrade_seqno(..) or something like that - does that sound right? going to continue reading

oh just saw this after i wrote the last response (which happened to include my initial thoughts on this idea, heh)

Reply:
Ok I thought about this overnight. Here's what makes sense to me (warning: rough spitball):

1. When a reader is instantiated its seqno since is `[current_max_seqno]`
2. It can downgrade its `seqno_since` to `[]` but then it loses the ability to perform any snapshots or subscriptions
3. When it calls `snapshot` it gets a `snapshot_seqno` that all batches within the snapshot are found in. That `snapshot_seqno` is >= the readers `seqno_since`. When the reader is done reading the snapshot it can downgrade its `seqno_since` to `snapshot_seqno`
4. Each batch that gets returned from a `subscribe` call has its own associated `batch_seqno` the stream of batch_seqno's are weakly increasing and all `batch_seqno`'s returned from subscribe have to be >= the readers current `seqno_since`. When a reader is done with a particular batch it can downgrade its `seqno_since` to `batch_seqno`
5. We can combine the time and seqno downgrades into one operation `downgrade_since_time_and_seqno`